### PR TITLE
[hip][cuda] Merge the tracing implementations.

### PR DESCRIPTION
**Pull request by AWoloszyn**
These were entirely copy-pasted off one another and it's not likely they will have to diverge in the future.

### CODE REVIEW

**Comment by AWoloszyn** [comment]
I currently have 2 different interfaces into the shared code 
`iree_hal_deferred_work_queue_device_interface_t` and `iree_hal_tracing_device_interface_t`. They are similar, and share some similar methods, but I left them as individual interfaces as they are nominally independent. But it means a bit of duplication of code. I was wondering if anyone else had an opinion on this.,

**Review by benvanik — CHANGES_REQUESTED** [review]
Let's try to pick another name besides `util/tracing.h` (`iree_hal_tracing_`) as this is only applicable to HIP and CUDA. We need a good name for that kind of thing - maybe just `stream_tracing.h` (`iree_hal_stream_tracing_`)?

It's fine to duplicate the interface as we want to keep the code decoupled and implementations may not use both. At some point I hope we delete the stream implementations entirely (or at least use them only as a debugging feature) in which case we may not even care about tracing there.

(still need to read through the rest of the PR, but wanted to flag naming early)

File: runtime/src/iree/hal/drivers/hip/rccl_channel.c [hunk 1]
```diff
@@ -592,11 +592,10 @@ iree_status_t iree_hal_hip_nccl_submit_batch(
     iree_hal_collective_batch_entry_t* entry = &batch->entries[i];
     iree_string_view_t collective_str =
         iree_hal_collective_op_format(&entry->op, &string_temp);
-    IREE_HIP_STREAM_TRACE_ZONE_BEGIN_EXTERNAL(
-        tracing_context, tracing_event_list, stream,
-        IREE_HAL_HIP_TRACING_VERBOSITY_FINE, __FILE__, strlen(__FILE__),
-        (uint32_t)__LINE__, __FUNCTION__, strlen(__FUNCTION__),
-        collective_str.data, collective_str.size);
+    IREE_STREAM_TRACE_ZONE_BEGIN_EXTERNAL(
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/hip/rccl_channel.c]:
```suggestion
    IREE_HAL_STREAM_TRACE_ZONE_BEGIN_EXTERNAL(
```
(and `IREE_HAL_GRAPH_TRACE_*` for the others)

**Review by benvanik — CHANGES_REQUESTED** [review]
Besides a naming nit on the macros LGTM once moved from utils/tracing to something else

**Review by benvanik — APPROVED** [review]

**Pull request merged**